cdap-proto/src/main/java/co/cask/cdap/proto/id/ProgramId.java

```diff
@@ -78,6 +79,12 @@ public String getEntityName() {
     return getProgram();
   }
 
+  @Override
+  public MetadataEntity toMetadataEntity() throws UnsupportedOperationException {
+    return MetadataEntity.ofNamespace(namespace).of(MetadataEntity.APPLICATION, application)
+      .of(MetadataEntity.TYPE, type.getPrettyName()).of(MetadataEntity.PROGRAM, program);
```

Comment:
not sure whether that makes sense? Should it be ```.of(type.getPrettyName(), program)```?

Reply:
also, for applications, you include the version in the MetadataEntity, but for programs you don't? What's the rationale behind that? Also, what does a version mean for metadata? Should meta data be version-independent? For example, if I deploy a new version of an app, should all the meta data now be reset? Or should it carry over? Does that mean we need a way to record version-specific and version-agnostic metadata?

Reply:
I think it will be nice to have constant and consistent keys making the programtype key might be confusing a little. 
the metadata version initial design is documented here : 
https://wiki.cask.co/display/CE/Metadata+Custom+Entities+and+Authorization